include/pisa/wand_data.hpp

```diff
@@ -25,72 +28,98 @@ class wand_data {
     wand_data(LengthsIterator len_it,
               uint64_t num_docs,
               binary_freq_collection const &coll,
+              std::string const &scorer_name,
               BlockSize block_size)
     {
         std::vector<uint32_t> doc_lens(num_docs);
+        uint32_t max_len = 0;
         std::vector<float> max_term_weight;
-        std::vector<uint32_t> terms_count;
+        std::vector<uint32_t> term_counts;
+        std::vector<uint32_t> term_lens;
         global_parameters params;
         double collection_len = 0;
         spdlog::info("Reading sizes...");
 
         for (size_t i = 0; i < num_docs; ++i) {
-            float len = *len_it++;
+            uint32_t len = *len_it++;
+            max_len = std::max(max_len, len);
             doc_lens[i] = len;
             collection_len += len;
         }
 
         float avg_len = float(collection_len / double(num_docs));
 
         typename block_wand_type::builder builder(coll, params);
+
+        for (auto const &seq : coll) {
+            size_t term_count = std::accumulate(seq.freqs.begin(), seq.freqs.end(), 0);
+            term_counts.push_back(term_count);
+            term_lens.push_back(seq.docs.size());
+        }
+        m_doc_lens.steal(doc_lens);
+        m_term_counts.steal(term_counts);
+        m_term_lens.steal(term_lens);
+        m_avg_len = avg_len;
+        m_max_len = max_len;
+        m_num_docs = num_docs;
+        m_collection_len = collection_len;
+
+        auto scorer_ptr = get_scorer(scorer_name, *this);
         {
             pisa::progress progress("Processing posting lists", coll.size());
+            size_t term_id = 0;
             for (auto const &seq : coll) {
-                size_t term_count = std::accumulate(seq.freqs.begin(), seq.freqs.end(), 0);
-                terms_count.push_back(term_count);
-                auto v = builder.add_sequence(seq, coll, doc_lens, avg_len, block_size);
+                auto v = builder.add_sequence(
+                    seq, coll, doc_lens, avg_len, (*scorer_ptr)(term_id), block_size);
```

Comment:
See, this is awkward, it would be much more clear if it was `scorer_ptr->term_scorer(term_id)`. Also, not sure we need that `_ptr`, this is what the type system is for. It's not like you have both pointer and object and need to differentiate.